File: sample_algorithms/Sorting/selectionsort/python/selectionsort.py
# Simple Selection Sort in Python
from tracers.tracer import Tracer
from helpers import swap  # Helper function: swaps arr[i] and arr[j] in-place

# [ALGORITHM]
def selection_sort(arr, tracer):
    n = len(arr)
    
    for i in range(n - 1):
        # Show the position where we'll place the minimum
        tracer.add_state(arr.copy(), selected=[i], indexVars=['i'], variables={'i': i})
        
        # Find the minimum element in the unsorted portion
        min_index = i
        for j in range(i + 1, n):
            tracer.add_state(arr.copy(), comparing=[min_index, j], selected=[i], indexVars=['i', 'j', 'min_index'], variables={'i': i, 'j': j, 'min_index': min_index})
            if arr[j] < arr[min_index]:
                min_index = j
        
        # If minIndex changed, update the selected position
        tracer.add_state(arr.copy(), selected=[i, min_index], indexVars=['i', 'min_index'], variables={'i': i, 'min_index': min_index})
        # Swap the minimum element with the first element of unsorted portion
        if min_index != i:
            swap(arr, i, min_index)
            tracer.add_state(arr.copy(), swapped=[i, min_index], selected=[i, min_index], indexVars=['i', 'min_index'], variables={'i': i, 'min_index': min_index})
    
    tracer.add_state(arr.copy()) # Complete state
    return arr, tracer

# [TEST]
if __name__ == "__main__":
    # [PARAMS]
    original_arr = [92, 14, 461, 1122, 235, 9, 127]
    # [/PARAMS]
    sorted_arr, tracer = selection_sort(original_arr.copy(), Tracer(category='sorting'))

    print(f'Original array: {original_arr}')
    print(f'Sorted array: {sorted_arr}')
    
    # Output tracer data for visualization
    tracer.finalize()

File: sample_algorithms/Sorting/selectionsort/javascript/selectionsort.js
import Tracer from './runtime/tracer.js';
import { swap } from './helpers.js';

function selectionSort(arr, tracer) {
    const n = arr.length;
    
    for (let i = 0; i < n - 1; i++) {
        tracer.add_state([...arr], { selected: [i], indexVars: ['i'], variables: { i } });
        
        let minIndex = i;
        
        // Find the minimum element in the unsorted portion
        for (let j = i + 1; j < n; j++) {
            tracer.add_state([...arr], { comparing: [minIndex, j], selected: [i], indexVars: ['i', 'j', 'minIndex'], variables: { i, j, minIndex } });  // Keep selected position visible
            if (arr[j] < arr[minIndex]) {
                minIndex = j;
            }
        }
        
        // If minIndex changed, update the selected position
        tracer.add_state([...arr], { comparing: [i, minIndex], indexVars: ['i', 'minIndex'], variables: { i, minIndex } });
        // Swap the minimum element with the first element of unsorted portion
        if (minIndex !== i) {
            swap(arr, i, minIndex);
            tracer.add_state([...arr], { swapped: [i, minIndex], selected: [i], indexVars: ['i', 'minIndex'], variables: { i, minIndex } });
        }
    }
    tracer.add_state([...arr]); // Complete state
    return [arr, tracer];
}

// [PARAMS]
const originalArr = [64, 34, 25, 12, 22, 11, 90];
// [/PARAMS]
const tracer = new Tracer('sorting'); // Tracer instance
const [sortedArr] = selectionSort([...originalArr], tracer);

console.log(`Original array: [${originalArr.join(", ")}]`);
console.log(`Sorted array: [${sortedArr.join(", ")}]`);

// Output tracer data for visualization
tracer.finalize();
